HIP/hip-513.md

```diff
@@ -0,0 +1,293 @@
+
+Proposes additional data placed in sidecar transaction records to enable full
+replay of transactions by mirror nodes and downstream users of Hedera.
+
+## Motivation
+
+Enabling mirror nodes to download traceability info only if they need or want to process it.
+
+Enabling full replay of smart contract service transactions will improve the
+usability, auditability, and debuggability of smart contract service
+transactions. Current facilities are insufficient to inspect "internal"
+transactions executed across contracts, and provide insufficient error messages
+when correcting errors during smart contract development.
+
+## Rationale
+
+[HIP-435](./hip-435.md) allows the creation of sidecar record files containing
+detailed information about transaction executions without bloating the main
+transaction record stream. This allows us to externalise verbose debug information
+about smart contract transactions that was previously expensive to download and
+process every time.
+
+## User stories
+
+Аs a mirror node operator, I want to have the option to download verbose contract operations-related data
+like contract state changes, contract actions, or contract bytecode.
+
+## Specification
+
+We will introduce a couple of sidecar record types, one containing the minimum
+needed data that allows a user to fully replay the transaction and derive the same
+set of results, and the other containing an entire operation-by-operation trace.
+
+### Add ContractStateChanges sidecar record type
+
+Add and populate a new sidecar record of type `ContractStateChanges`. It will
+have a single field - repeated `ContractStateChange`, which will be populated
+whenever a transaction causes contract storage states to change.
+This field will include the changes across all contract storage and not just the
+initial contract in the transaction.
+
+```
+message ContractStateChanges {
+  repeated ContractStateChange contract_state_changes = 1;
+}
+```
+
+For each slot read or written by the smart contract transaction
+a `ContactStateChange` record should be added to the `ContractStateChanges`.
+
+### Add ContractAction sidecar record type
+
+Add and populate a new sidecar record of type `ContractAction`. It will
+have fine-grained information about transaction. Which will improve debugging
+and auditing.
+
+```
+/**
+ * The type of action described by the action proto.
+ */
+enum ContractActionType {
+    /**
+     * default non-value.
+     */
+    NoAction = 0;
+
+    /**
+     * Most CALL, CALLCODE, DELEGATECALL, and STATICCALL, and first action of ContractCall/ContractCallLocal to deployed
+     * contracts. This does not include calls to system or precompiled contracts.
+     */
+    Call = 1;
+
+    /**
+     * CREATE, CREATE2, and first action of ContractCreate.
+     */
+
+    Create = 2;
+
+    /**
+     * like Call, but to precompiled contracts (0x1 to 0x9 as of Berlin)
+     */
+    Precompile = 3;
```

Comment:
Precompiles show up in the record stream as a separate HAPI transaction with a unique consensus timestamp. Will the contract actions for a precompile be associated with the parent contractcall/contractcreate/ethereumtransaction timestamp or with its child precompile transaction timestamp?

Reply:
Ok, so HTS precompiled contracts won't show up separately in the sidecar but will always be attached to a top level HAPI transaction list of actions? That makes sense to me. Might be good to clarify what sorts of transactions will have actions in the sidecar in the HIP.

Reply:
I would prefer that precompiles still create contract actions in the sidecars.  (a) we would have to detect the system contract and _not_ emit a contract action and (b) if we took it out we would need to "swizzle" the data back in for a trace of the call.

I would rather not take effort to take it out and put it back in to save data in the sidecar.

Reply:
Precompiles will create `ContractActions` in _a_ sidecar record, the question is what timestamp that sidecar should have (which transaction will the sidecar be associated with), the top-level tx, or the child record?

Reply:
For example, would it make sense to have a Child record of type `CryptoApprove` for some HTS precompile invocation, and have a sidecar record with `ContractActions` that is associated with the child record instead of the parent ContractCall transaction

Reply:
As agreed on the parking lot yesterday, we will create `ContractAction` entries and use the top-level record timestamp in the `TransactionSidecarRecord` where we list them.